src/stream/src/executor/source/mod.rs

```diff
@@ -133,28 +133,49 @@ pub async fn apply_rate_limit(stream: BoxChunkSourceStream, rate_limit_rps: Opti
         )
     });
 
+    fn compute_chunk_permits(chunk: &StreamChunk, limit: usize) -> usize {
+        let chunk_size = chunk.capacity();
+        let ends_with_update = if chunk_size >= 2 {
+            // Note we have to check if the 2nd last is `U-` to be consistenct with `StreamChunkBuilder`.
+            // If something inconsistent happens in the stream, we may not have `U+` after this `U-`.
+            chunk.ops()[chunk_size - 2].is_update_delete()
+        } else {
+            false
+        };
+        if chunk_size == limit + 1 && ends_with_update {
+            // If the chunk size exceed limit because of the last `Update` operation,
+            // we should minus 1 to make sure the permits consumed is within the limit (max burst).
+            chunk_size - 1
+        } else {
+            chunk_size
+        }
+    }
+
     #[for_await]
     for chunk in stream {
         let chunk = chunk?;
-        let cardinality = chunk.cardinality();
+        let chunk_size = chunk.capacity();
 
-        if rate_limit_rps.is_none() || cardinality == 0 {
+        if rate_limit_rps.is_none() || chunk_size == 0 {
             // no limit, or empty chunk
             yield chunk;
             continue;
         }
 
         let limiter = limiter.as_ref().unwrap();
-        let limit = rate_limit_rps.unwrap();
-        if cardinality <= limit as usize {
-            let n = NonZeroU32::new(cardinality as u32).unwrap();
+        let limit = rate_limit_rps.unwrap() as usize;
+
+        let chunk_size = compute_chunk_permits(&chunk, limit);
+        if chunk_size <= limit {
```

Comment:
We don't need to call `compute_chunk_permits` here.
If there's a `U+ / U-` chunk which can't be split, it will fallback to the `else` block.

This can just be:
```
if chunk.size() <= limit {
```

Reply:
If don't do `compute_chunk_permits` before the first check, we will do some unnecessary work in `chunk.split()` and get the exact same chunk again. Since `compute_chunk_permits` is actually much cheaper than `split()`, I guess we'd better check it first. Another approach is to check the ending `Update` in `chunk.split()`, which will potentially benefit more calling spots, but with an overhead of `Vec` construction. Either LGTM.

Reply:
Hmm I guess it's fine also. Not too concerned about cost here, since bottleneck will be rate limit.

Btw you may rename `chunk_size` here to `required_permits` instead. I think with that change it is a lot more readable.